lib/std/priority_dequeue.zig

```diff
@@ -0,0 +1,879 @@
+            }
+        }
+
+        fn bestItemAtIndices(self: Self, index1: usize, index2: usize, compare: fn (Self, T, T) bool) ItemAndIndex {
+            var item1 = self.getItem(index1);
+            var item2 = self.getItem(index2);
+            return self.bestItem(item1, item2, compare);
+        }
+
+        fn bestDescendent(self: Self, first_child_index: usize, first_grandchild_index: usize, compare: fn (Self, T, T) bool) ItemAndIndex {
+            const second_child_index = first_child_index + 1;
+            if (first_grandchild_index >= self.len) {
+                // No grandchildren, find the best child (second may not exist)
+                if (second_child_index >= self.len) {
+                    return .{
+                        .item = self.items[first_child_index],
+                        .index = first_child_index,
+                    };
+                } else {
+                    return self.bestItemAtIndices(first_child_index, second_child_index, compare);
+                }
+            }
+
+            const second_grandchild_index = first_grandchild_index + 1;
+            if (second_grandchild_index >= self.len) {
+                // One grandchild, so we know there is a second child. Compare first grandchild and second child
+                return self.bestItemAtIndices(first_grandchild_index, second_child_index, compare);
+            }
+
+            const best_left_grandchild_index = self.bestItemAtIndices(first_grandchild_index, second_grandchild_index, compare).index;
+            const third_grandchild_index = second_grandchild_index + 1;
+            if (third_grandchild_index >= self.len) {
+                // Two grandchildren, and we know the best. Compare this to second child.
+                return self.bestItemAtIndices(best_left_grandchild_index, second_child_index, compare);
+            } else {
+                // Three grandchildren, compare the min of the first two with the third
+                return self.bestItemAtIndices(best_left_grandchild_index, third_grandchild_index, compare);
+            }
+        }
+
+        /// Return the number of elements remaining in the heap
+        pub fn count(self: Self) usize {
+            return self.len;
+        }
+
+        /// Return the number of elements that can be added to the
+        /// dequeue before more memory is allocated.
+        pub fn capacity(self: Self) usize {
+            return self.items.len;
+        }
+
+        /// Heap takes ownership of the passed in slice. The slice must have been
+        /// allocated with `allocator`.
+        /// De-initialize with `deinit`.
+        pub fn fromOwnedSlice(allocator: *Allocator, lessThanFn: fn (T, T) bool, items: []T) Self {
+            var dequeue = Self{
+                .items = items,
+                .len = items.len,
+                .allocator = allocator,
+                .lessThanFn = lessThanFn,
+            };
+            const half = (dequeue.len >> 1) - 1;
```

Comment:
Underflow if `dequeue` has 1 or 0 elements.

Reply:
Thanks, fixed here and in `std.PriorityQueue` (and tests added).